### DISCUSSION TITLE
Scope state is ignored when logging to the Console

### DISCUSSION THREAD

**Discussion by fgheysels** [arcus-azure/arcus.observability#204] [Q&A] [answered]
Consider the following code:

```csharp
 using (_logger.BeginScope(new Dictionary<string, object>()
{
    ["Item"] = "foo",
    ["Error"] = "bar"
}))
{
      _logger.LogError("An error occured!");
}
```

When logging to the console, I'd expect that the additional properties 'item' and 'error' are also visible in the console-logs.  This is however not the case:

![image](https://user-images.githubusercontent.com/3605786/115291543-2c275780-a155-11eb-80de-d9b55142d189.png)

Simple repro-sample:

```csharp
class Program
{
    static async Task Main(string[] args)
    {
         var host = Host.CreateDefaultBuilder(args)
                           .ConfigureLogging(logging => logging.AddConsole())
                           .ConfigureServices(services =>
                           {
                               services.AddHostedService<TestClass>();
                           })
                           .UseSerilog(ConfigureLogger)
                           .UseConsoleLifetime()
                           .Build();

            await host.RunAsync();
    }

    private static void ConfigureLogger(HostBuilderContext context, LoggerConfiguration loggerConfiguration)
    {
            loggerConfiguration.MinimumLevel.Information()
                               .Enrich.FromLogContext()
                               .WriteTo.Console();
    }
}

public class TestClass : BackgroundService
{
    private readonly ILogger<TestClass> _logger;

    public TestClass(ILogger<TestClass> logger)
    {
          _logger = logger;
    }

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
            while (stoppingToken.IsCancellationRequested == false)
            {
                using (_logger.BeginScope(new Dictionary<string, object>()
                {
                    ["Item"] = "foo",
                    ["Error"] = "bar"
                }))
               {
                    _logger.LogError("An error occured!");
               }

                await Task.Delay(500);
           }
    }
}
```

**Comment by stijnmoreels** [comment]
Hmm, I don't think this has anything to do with Arcus, as both `.LogError` and `.BeginScope` are concepts outside Arcus. The telemetry context you referring to is only available for Arcus telemetry extensions that actually uses this context. The Serilog extensions doesn't know anything about this.

**Comment by tomkerkhove** [comment]
This is the native console sink for Serilog indeed, but you can change the log output template to surface this, see https://github.com/arcus-azure/arcus.observability/issues/185#issuecomment-792539786 for an example

**Comment by fgheysels** [comment, marked as answer]
Indeed, makes sense.
Also, thanks @tomkerkhove  for pointing me to my other issue which I've absolutely forgotten about :)

The solution is indeed to expand the output-template of the Serilogger.  It is even documented on how to get those extra properties / dimensions into the output.   The [link](https://github.com/serilog/serilog/wiki/Formatting-Output) you've posted in the other issue gives the solution.   

The properties can be visualized by the logger by specifying the `Properties` keyword in the output-template like this:

```csharp
loggerConfiguration.MinimumLevel.Information()
                   .WriteTo.Console(outputTemplate: "[{Timestamp}] {Level} {Message} {Properties:j}{NewLine}");
```

**Comment by tomkerkhove** [comment]
@fgheysels In terms of our templates, would you like or dislike if that was enabled out-of-the-box?

This has cost impact since logs become better but can be nice addition.

**Reply by fgheysels** [reply]
I think it'd be nice if this is enabled by default.  
The cost impact is that logs are becoming bigger ?

**Reply by tomkerkhove** [reply]
Yes. Console logs for apps on AKS will go to Log Analytics for example.

**Reply by fgheysels** [reply]
On the other hand, what if you don't do this in the template, but some comments are added before the line where the outputTemplate is set, with a referal to the Serilog docs that point out the possible keywords ?

**Reply by tomkerkhove** [reply]
I think that sounds reasonable and we can even add the template as a comment so people have an example.

Thoughts @stijnmoreels?

**Reply by stijnmoreels** [reply]
I think it's valuable to guide consumers towards implementations (like we guide them towards our Key Vault secret provider). But on the other hand, we should be careful not to overload our templates with lots of comments and tips/tricks which will most of them be removed after the consumer has created a project from the template (I think). Maybe some of these comments can be added to our documentation page instead.

**Reply by tomkerkhove** [reply]
I do get that, but with a template I don't think people will notice that so for me the options are:
1. Using it by default, knowing it increases cost, or
2. Keep using the current approach, but add a comment

**Reply by fgheysels** [reply]
Actually, we can do both as well :)

I think I would definitely add a comment so that users of the template know where to find some additional information when they want to change their default template.  It frees them from doing some research themselves which can be time consuming.

Also a good idea from Stijn to put some guidance in the documents, but ......   I think we all know that few people actually read the documentation.  Having it in a comment immediately, gives some more visibility imo.

**Reply by tomkerkhove** [reply]
Fine by me

**Reply by stijnmoreels** [reply]
Hmm, maybe a recap of the conclusion of this discussion is needed for us (read: me 😄 ) to know what output template we should use instead in our project templates.
Also, keep in mind, with the changed message template that was requested (to be serializable to JSON) the context properties are not available directly anymore as a log property.
